README.md
# ringtone_set_mul

Flutter package for easily setting device's default ringtone, notification sound and alarm.

> [!IMPORTANT]
> Works only on Android


original rep:

[GitHub - rtacr/ringtone_set: Flutter package for setting the default ringtone or notification sound](https://github.com/rtacr/ringtone_set)



## Getting Started

Add these lines to **AndroidManifest.xml**. 

```xml
<manifest>
...
    <uses-permission android:name="android.permission.WRITE_SETTINGS"
        tools:ignore="ProtectedPermissions" />
    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
...
</manifest>
```

### Set from assets

Put your sounds into \<yourapp\>/assets/
add them to your **pubspec.yaml**

```yaml
flutter:
  assets:
    - assets/music.mp3
```

```yaml
dependencies:
  ringtone_set_mul: ^0.0.1
```

Import the package

```Dart
import 'package:ringtone_set_mul/ringtone_set_mul.dart.dart';
```

Call the function

```Dart
RingtoneSet.setRingtone("assets/music.mp3");
```

or

```Dart
RingtoneSet.setNotification("assets/music.mp3");
```

or

```Dart
RingtoneSet.setAlarm("assets/music.mp3");
```

### Set from Network

You can also set ringtone from network URL:

```Dart
RingtoneSet.setRingtoneFromNetwork("https://example.com/music.mp3");
```

or

```Dart
RingtoneSet.setNotificationFromNetwork("https://example.com/music.mp3");
```

or

```Dart
RingtoneSet.setAlarmFromNetwork("https://example.com/music.mp3");
```

### Set from File

Finally, you can set ringtone from a `File` object as well:

```Dart
final File ringtoneFile = File("music.mp3");

RingtoneSet.setRingtoneFromFile(ringtoneFile);

RingtoneSet.setNotificationFromFile(ringtoneFile);

RingtoneSet.setAlarmFromFile(ringtoneFile);
```


CHANGELOG.md
## 0.0.1


* Initial Release

## 0.0.2
- fix dep
## 0.0.3
-
## 0.0.4
- 新增判断
- ## 0.0.5
- remove ndk 

test/ringtone_set_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:ringtone_set_mul/ringtone_set_mul.dart.dart';

void main() {
  const MethodChannel channel = MethodChannel('ringtone_set');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await RingtoneSet.platformVersion, '42');
  });
}


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:ringtone_set_mul/ringtone_set_mul.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {}

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Builder(builder: (context) {
          return Center(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: <Widget>[
                TextButton(
                  onPressed: () async {
                    bool success = false;
                    try {
                      success = await RingtoneSet.setMultipleAsset(
                          "assets/h2.mp3",
                          isAlarm: true,
                          isNotification: true,
                          isRingtone: true);
                    } on PlatformException {
                      success = false;
                    }
                    var snackBar;
                    if (success) {
                      snackBar = SnackBar(
                        content: Text("Multiple set successfully!"),
                      );
                    } else {
                      snackBar = SnackBar(content: Text("Error"));
                    }
                    ScaffoldMessenger.of(context).showSnackBar(snackBar);
                  },
                  child: Text("Hey Multiple"),
                ),
                TextButton(
                  onPressed: () async {
                    bool success = false;
                    try {
                      success = await RingtoneSet.setRingtone("assets/h1.mp3");
                    } on PlatformException {
                      success = false;
                    }
                    var snackBar;
                    if (success) {
                      snackBar = SnackBar(
                        content: Text("Ringtone set successfully!"),
                      );
                    } else {
                      snackBar = SnackBar(content: Text("Error"));
                    }
                    ScaffoldMessenger.of(context).showSnackBar(snackBar);
                  },
                  child: Text("Hey Ringtone"),
                ),
                TextButton(
                  onPressed: () async {
                    bool success = false;
                    try {
                      success = await RingtoneSet.setNotification(
                        "assets/h1.mp3",
                      );
                    } on PlatformException {
                      success = false;
                    }
                    var snackBar;
                    if (success) {
                      snackBar = SnackBar(
                        content: Text("Notification set successfully!"),
                      );
                    } else {
                      snackBar = SnackBar(content: Text("Error"));
                    }
                    ScaffoldMessenger.of(context).showSnackBar(snackBar);
                  },
                  child: Text("Hey Notification"),
                ),
                TextButton(
                  onPressed: () async {
                    bool success = false;
                    try {
                      success = await RingtoneSet.setRingtone(
                        "assets/untitled.mp3",
                      );
                    } on PlatformException {
                      success = false;
                    }
                    var snackBar;
                    if (success) {
                      snackBar =
                          SnackBar(content: Text("Ringtone set successfully!"));
                    } else {
                      snackBar = SnackBar(content: Text("Error"));
                    }
                    ScaffoldMessenger.of(context).showSnackBar(snackBar);
                  },
                  child: Text("Untitled Ringtone"),
                ),
                TextButton(
                  onPressed: () async {
                    bool success = false;
                    try {
                      success = await RingtoneSet.setNotification(
                        "assets/untitled.mp3",
                      );
                    } on PlatformException {
                      success = false;
                    }
                    var snackBar;
                    if (success) {
                      snackBar = SnackBar(
                        content: Text("Notification set successfully!"),
                      );
                    } else {
                      snackBar = SnackBar(content: Text("Error"));
                    }
                    ScaffoldMessenger.of(context).showSnackBar(snackBar);
                  },
                  child: Text("Untitled Notification"),
                ),
              ],
            ),
          );
        }),
      ),
    );
  }
}


lib/ringtone_set.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/services.dart';
import 'package:ringtone_set/src/setter_functions.dart';

/// A class for setting ringtones, notifications, alarms.
class RingtoneSet {
  static const MethodChannel _channel = const MethodChannel('ringtone_set');

  /// Android version.
  static Future<String> get platformVersion async {
    final String version = await _channel.invokeMethod('getPlatformVersion');
    return version;
  }

  /// Android SDK version code.
  static Future<int> get platformSdk async {
    final int sdk = await _channel.invokeMethod('getPlatformSdk');
    return sdk;
  }

  ///多选!!!
  static Future<bool> setMultipleFile(File file,
      {bool isNotification = false,
      bool isRingtone = false,
      bool isAlarm = false}) async {
    return setMultipleFromFile(
        file: file,
        action: 'setMultiple',
        isNotification: isNotification,
        isRingtone: isRingtone,
        isAlarm: isAlarm);
  }

  ///多选
  static Future<bool> setMultipleAsset(String asset,
      {bool isNotification = false,
      bool isRingtone = false,
      bool isAlarm = false}) async {
    return setMultipleFromAsset(
        asset: asset,
        action: 'setMultiple',
        isNotification: isNotification,
        isRingtone: isRingtone,
        isAlarm: isAlarm);
  }

  /// Status of the `WRITE_SETTINGS` permission.
  ///
  /// Returns `true` if the permission has been granted.
  /// On Android 5.1(SDK 22) and older always returns `true`.
  static Future<bool> get isWriteSettingsGranted async {
    final bool granted = await _channel.invokeMethod('isWriteGranted');
    return granted;
  }

  /// Sets ringtone from asset.
  static Future<bool> setRingtone(String asset) async {
    return setFromAsset(
      asset: asset,
      action: 'setRingtone',
    );
  }

  /// Sets ringtone from network URL.
  static Future<bool> setRingtoneFromNetwork(String url) async {
    return setFromNetwork(
      url: url,
      action: 'setRingtone',
    );
  }

  /// Sets ringtone from file.
  static Future<bool> setRingtoneFromFile(File file) async {
    return setFromFile(
      file: file,
      action: 'setRingtone',
    );
  }

  /// Sets notification from asset path.
  static Future<bool> setNotification(String asset) async {
    return setFromAsset(
      asset: asset,
      action: 'setNotification',
    );
  }

  /// Sets notification from network URL.
  static Future<bool> setNotificationFromNetwork(String url) async {
    return setFromNetwork(
      url: url,
      action: 'setNotification',
    );
  }

  /// Sets notification from file.
  static Future<bool> setNotificationFromFile(File file) async {
    return setFromFile(
      file: file,
      action: 'setNotification',
    );
  }

  /// Sets alarm from asset path.
  static Future<bool> setAlarm(String asset) async {
    return setFromAsset(
      asset: asset,
      action: 'setAlarm',
    );
  }

  /// Sets alarm from network URL.
  static Future<bool> setAlarmFromNetwork(String url) async {
    return setFromNetwork(
      url: url,
      action: 'setAlarm',
    );
  }

  /// Sets alarm from file.
  static Future<bool> setAlarmFromFile(File file) async {
    return setFromFile(
      file: file,
      action: 'setAlarm',
    );
  }
}


lib/src/setter_functions.dart
import 'dart:io';

import 'package:flutter/services.dart';
import 'package:http/http.dart' as http;
import 'package:path_provider/path_provider.dart';
import 'package:ringtone_set_mul/ringtone_set_mul.dart.dart';

const MethodChannel _channel = const MethodChannel('ringtone_set');

/// Sets [action] from asset path.
///
/// [action] can be `"setRingtone"`, `"setNotification"`, `"setAlarm"`.
Future<bool> setFromAsset({
  required String asset,
  required String action,
}) async {
  final path = await _getPath(
    src: asset,
    storageDirectoryType: _getStorageDirectoryType(action),
  );
  final file = File(path);
  final loadedAsset = await rootBundle.load(asset);
  await file.writeAsBytes((loadedAsset).buffer.asUint8List());

  final bool result = await _channel.invokeMethod(action, {"path": file.path});
  return result;
}

/// Sets [action] from network URL.
///
/// [action] can be `"setRingtone"`, `"setNotification"`, `"setAlarm"`.
Future<bool> setFromNetwork({
  required String url,
  required String action,
}) async {
  final path = await _getPath(
    src: url,
    storageDirectoryType: _getStorageDirectoryType(action),
  );
  final file = File(path);
  final response = await http.get(Uri.parse(url));
  if (response.statusCode == 200) {
    await file.writeAsBytes(response.bodyBytes);
    final mimeType = await _parseMimeType(response);

    final bool result = await _channel.invokeMethod(
      action,
      {"path": file.path, "mimeType": mimeType},
    );

    return result;
  } else {
    throw Exception(
      "Network error. Code ${response.statusCode}, ${response.reasonPhrase}",
    );
  }
}

/// Returns path for storing ringtone/notification/alarm sound.
///
/// On Android 10 or newer, the path is in the temporary directory.
/// On older Android versions, the path is in one of the external storage
/// directories.
///
/// [src] can be an asset path or a URL.
///
/// [storageDirectoryType] should be [StorageDirectory.ringtones],
/// [StorageDirectory.notifications], [StorageDirectory.alarms].
Future<String> _getPath({
  required String src,
  required StorageDirectory? storageDirectoryType,
}) async {
  final int sdk = await _channel
      .invokeMethod('getPlatformSdk'); //await RingtoneSet.platformSdk;

  if (sdk >= 29) {
    // Android 10 or newer
    return '${(await getTemporaryDirectory()).path}/${src.split('/').last}';
  } else {
    // Android 9 or older
    final ringtoneDirectories = await getExternalStorageDirectories(
      type: storageDirectoryType,
    );
    final directory = ringtoneDirectories!.first.path;
    return '$directory/${src.split('/').last}';
  }
}

/// Returns the temporary directory type according to action.
///
/// [action] can be `"setRingtone"`, `"setNotification"`, `"setAlarm"`.
StorageDirectory? _getStorageDirectoryType(String action) {
  if (action == "setRingtone") {
    return StorageDirectory.ringtones;
  }
  if (action == "setNotification") {
    return StorageDirectory.notifications;
  }
  if (action == "setAlarm") {
    return StorageDirectory.alarms;
  }
  return null;
  // throw Exception("Unknown set action!");
}

/// Parses mimeType from the "content-type" header of [response].
Future<String?> _parseMimeType(http.Response response) async {
  final headers = response.headers;
  final contentTypeHeader = headers["content-type"];
  String? mimeType;
  if (contentTypeHeader != null) {
    final contentType = ContentType.parse(contentTypeHeader);
    mimeType = contentType.mimeType;
  }
  return mimeType;
}

/// Sets [action] from file.
///
/// [action] can be `"setRingtone"`, `"setNotification"`, `"setAlarm"`.
Future<bool> setFromFile({
  required File file,
  required String action,
}) async {
  final bool result = await _channel.invokeMethod(action, {"path": file.path});

  return result;
}

Future<bool> setMultipleFromFile(
    {required File file,
    required String action,
    required bool isNotification,
    required bool isRingtone,
    required bool isAlarm})
async {
  final bool result = await _channel.invokeMethod(action, {
    "path": file.path,
    "isRingt": isRingtone,
    "isNotif": isNotification,
    "isAlarm": isAlarm
  });
  return result;
}

Future<bool> setMultipleFromAsset(
    {required String asset,
    required String action,
    required bool isNotification,
    required bool isRingtone,
    required bool isAlarm}) async {
  final path = await _getPath(
    src: asset,
    storageDirectoryType: _getStorageDirectoryType(action),
  );
  final file = File(path);
  final loadedAsset = await rootBundle.load(asset);
  await file.writeAsBytes((loadedAsset).buffer.asUint8List());

  final bool result = await _channel.invokeMethod(action, {
    "path": file.path,
    "isRingt": isRingtone,
    "isNotif": isNotification,
    "isAlarm": isAlarm
  });
  return result;
}


android/src/main/java/acr/rt/ringtone_set_mul/RingtoneSetPlugin.java
package android.src.main.java.acr.rt.ringtone_set_mul;

import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.media.RingtoneManager;
import android.net.Uri;
import android.os.Build;
import android.provider.MediaStore;
import android.provider.Settings;
import android.webkit.MimeTypeMap;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import java.io.BufferedInputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/**
 * RingtoneSetPlugin
 */
public class RingtoneSetPlugin implements FlutterPlugin, MethodCallHandler {
    private static RingtoneSetPlugin instance;
    private MethodChannel channel;
    private Context mContext;

    public static void registerWith(Registrar registrar) {
        if (instance == null) {
            instance = new RingtoneSetPlugin();
        }
        instance.onAttachedToEngine(registrar.context(), registrar.messenger());
    }

    @Override
    public void onAttachedToEngine(FlutterPluginBinding binding) {
        onAttachedToEngine(binding.getApplicationContext(), binding.getBinaryMessenger());
    }

    public void onAttachedToEngine(Context applicationContext, BinaryMessenger messenger) {
        if (channel != null) {
            return;
        }
        this.mContext = applicationContext;

        channel =
                new MethodChannel(
                        messenger, "ringtone_set");

        channel.setMethodCallHandler(this);
    }

    private boolean isSystemWritePermissionGranted() {
        boolean retVal = true;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            retVal = Settings.System.canWrite(mContext);
        }
        return retVal;
    }

    private void requestSystemWritePermission() {
        boolean retVal = isSystemWritePermissionGranted();
        if (!retVal) {
            if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
                Intent intent = new Intent(android.provider.Settings.ACTION_MANAGE_WRITE_SETTINGS);
                String both = "package:" + mContext.getPackageName();
                intent.setData(Uri.parse(both));
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                mContext.startActivity(intent);
            }
        }
    }

    private static String guessMimeTypeFromExtension(String path) {
        String mimeType = null;
        String fileExtension = "";
        try {
            int i = path.lastIndexOf('.');
            if (i > 0) {
                fileExtension = path.substring(i + 1);
            }
            if (fileExtension != "") {
                mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(fileExtension);
            }
        } catch (Exception ignored) {

        }

        return mimeType;
    }

    static private String guessMimeTypeFromStream(InputStream is)
            throws IOException {
        // If we can't read ahead safely, just give up on guessing
        if (!is.markSupported())
            return null;

        is.mark(16);
        int c1 = is.read();
        int c2 = is.read();
        int c3 = is.read();
        int c4 = is.read();
        int c5 = is.read();
        int c6 = is.read();
        int c7 = is.read();
        int c8 = is.read();
        int c9 = is.read();
        int c10 = is.read();
        int c11 = is.read();
        is.reset();

        if (c1 == 0x2E && c2 == 0x73 && c3 == 0x6E && c4 == 0x64) {
            return "audio/basic";  // .au format, big endian
        }

        if (c1 == 0x64 && c2 == 0x6E && c3 == 0x73 && c4 == 0x2E) {
            return "audio/basic";  // .au format, little endian
        }

        if (c1 == 'R' && c2 == 'I' && c3 == 'F' && c4 == 'F') {
            return "audio/x-wav";
        }

        if (c1 == 0x23 && c2 == 0x21 && c3 == 0x41 && c4 == 0x4D && c5 == 0x52) {
            return "audio/amr";
        }

        if (c1 == 0x66 && c2 == 0x74 && c3 == 0x79 && c4 == 0x70 && c5 == 0x69 && c6 == 0x73
                && c7 == 0x6F && c8 == 0x6D) {
            return "audio/mp4";
        }

        if (c1 == 0x00 && c2 == 0x00 && c3 == 0x00 && c4 == 0x20 && c5 == 0x66 && c6 == 0x74
                && c7 == 0x79 && c8 == 0x70 && c9 == 0x4D && c10 == 0x34 && c11 == 0x41) {
            return "audio/mp4";
        }

        if (c1 == 0x4D && c2 == 54 && c3 == 68 && c4 == 64) {
            return "audio/midi";
        }

        if (c1 == 0x1A && c2 == 0x45 && c3 == 0xDF && c4 == 0xA3) {
            return "audio/x-matroska";
        }

        if (c1 == 0x30 && c2 == 0x26 && c3 == 0xB2 && c4 == 0x75 && c5 == 0x8E && c6 == 0x66
                && c7 == 0xCF && c8 == 0x11) {
            return "audio/x-ms-wma";
        }

        if (c1 == 0x4F && c2 == 0x67 && c3 == 0x67 && c4 == 0x53) {
            return "audio/ogg";
        }

        if (c1 == 0x49 && c2 == 0x44 && c3 == 0x33) {
            return "audio/mpeg";
        }

        if (c1 == 0xFF && (c2 == 0xFB || c2 == 0xF3 || c2 == 0xF2)) {
            return "audio/mpeg";
        }

        return null;
    }

    private static String getMIMEType(String absolutePath, String downloadedMimeType) {
        String mimeType = null;
        mimeType = guessMimeTypeFromExtension(absolutePath);

        if (mimeType != null) {
            return mimeType;
        }

        try (InputStream inputStream = new BufferedInputStream(new FileInputStream(absolutePath))) {
            mimeType = guessMimeTypeFromStream(inputStream);
        } catch (Exception e) {
            e.printStackTrace();
        }

        if (mimeType != null) {
            return mimeType;
        }

        // downloadedMimeType obtained from "content-type" header
        if (downloadedMimeType != null) {
            // A little fix for a common server misconfiguration
            if (downloadedMimeType == "audio/wav" ) {
                return "audio/x-wav";
            }
            return downloadedMimeType;
        }

        return "audio/mpeg";
    }

    @RequiresApi(api = Build.VERSION_CODES.KITKAT)
    private void setThings(String path, String downloadedMimeType, boolean isRingt, boolean isNotif, boolean isAlarm) {
        requestSystemWritePermission();
        String s = path;
        File mFile = new File(s);  // set File from path
        if (mFile.exists()) {
            // Android 10 or newer
            if (android.os.Build.VERSION.SDK_INT > 28) {// file.exists
                ContentValues values = new ContentValues();
                values.put(MediaStore.MediaColumns.DATA, mFile.getAbsolutePath());
                values.put(MediaStore.MediaColumns.TITLE, "Custom ringtone");
                values.put(MediaStore.MediaColumns.MIME_TYPE, getMIMEType(mFile.getAbsolutePath(), downloadedMimeType));
                values.put(MediaStore.MediaColumns.SIZE, mFile.length());
                values.put(MediaStore.Audio.Media.ARTIST, "Ringtone app");
                values.put(MediaStore.Audio.Media.IS_RINGTONE, isRingt);
                values.put(MediaStore.Audio.Media.IS_NOTIFICATION, isNotif);
                values.put(MediaStore.Audio.Media.IS_ALARM, isAlarm);
                values.put(MediaStore.Audio.Media.IS_MUSIC, false);

                Uri newUri = mContext.getContentResolver().insert(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, values);

                try (OutputStream os = mContext.getContentResolver().openOutputStream(newUri)) {
                    int size = (int) mFile.length();
                    byte[] bytes = new byte[size];
                    try {
                        BufferedInputStream buf = new BufferedInputStream(new FileInputStream(mFile));
                        buf.read(bytes, 0, bytes.length);
                        buf.close();

                        os.write(bytes);
                        os.close();
                        os.flush();
                    } catch (IOException e) {
                    }
                } catch (Exception ignored) {
                    ignored.printStackTrace();
                }
                if (isNotif) {
                    RingtoneManager.setActualDefaultRingtoneUri(
                            mContext, RingtoneManager.TYPE_NOTIFICATION,
                            newUri);
                }
                if (isRingt) {
                    RingtoneManager.setActualDefaultRingtoneUri(
                            mContext, RingtoneManager.TYPE_RINGTONE,
                            newUri);
                }
                if (isAlarm) {
                    RingtoneManager.setActualDefaultRingtoneUri(
                            mContext, RingtoneManager.TYPE_ALARM,
                            newUri);
                }
            } else {
                // Android 9 or older
                final String absolutePath = mFile.getAbsolutePath();

                ContentValues values = new ContentValues();
                values.put(MediaStore.MediaColumns.DATA, absolutePath);
                values.put(MediaStore.MediaColumns.TITLE, "Custom ringtone");
                values.put(MediaStore.MediaColumns.SIZE, mFile.length());
                values.put(MediaStore.Audio.Media.ARTIST, "Ringtone app");
                values.put(MediaStore.Audio.Media.IS_RINGTONE, isRingt);
                values.put(MediaStore.Audio.Media.IS_NOTIFICATION, isNotif);
                values.put(MediaStore.Audio.Media.IS_ALARM, isAlarm);
                values.put(MediaStore.Audio.Media.IS_MUSIC, false);

                // insert it into the database
                Uri uri = MediaStore.Audio.Media.getContentUriForPath(absolutePath);

                // delete the old one first
                mContext.getContentResolver().delete(uri, MediaStore.MediaColumns.DATA + "=\"" + absolutePath + "\"", null);

                // insert a new record
                Uri newUri = mContext.getContentResolver().insert(uri, values);

                if (isNotif) {
                    RingtoneManager.setActualDefaultRingtoneUri(
                            mContext, RingtoneManager.TYPE_NOTIFICATION,
                            newUri);
                }
                if (isRingt) {
                    RingtoneManager.setActualDefaultRingtoneUri(
                            mContext, RingtoneManager.TYPE_RINGTONE,
                            newUri);
                }
                if (isAlarm) {
                    RingtoneManager.setActualDefaultRingtoneUri(
                            mContext, RingtoneManager.TYPE_ALARM,
                            newUri);
                }
            }
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.KITKAT)
    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
        if (call.method.equals("getPlatformVersion")) {
            result.success("Android " + android.os.Build.VERSION.RELEASE);
            return;
        } else if (call.method.equals("getPlatformSdk")) {
            result.success(android.os.Build.VERSION.SDK_INT);
            return;
        }else if (call.method.equals("setRingtone")) {
            String path = call.argument("path");
            String downloadedMimeType = call.argument("mimeType");
            setThings(path, downloadedMimeType, true, false, false);

            result.success(true);
            return;
        } else if (call.method.equals("setNotification")) {
            String path = call.argument("path");
            String downloadedMimeType = call.argument("mimeType");
            setThings(path, downloadedMimeType, false, true, false);

            result.success(true);
            return;
        } else if (call.method.equals("setAlarm")) {
            String path = call.argument("path");
            String downloadedMimeType = call.argument("mimeType");
            setThings(path, downloadedMimeType, false, false, true);

            result.success(true);
            return;
        } else if (call.method.equals("setMultiple")) {
            String path = call.argument("path");
            String downloadedMimeType = call.argument("mimeType");
            boolean isRingt = call.argument("isRingt");
            boolean isNotif = call.argument("isNotif");
            boolean isAlarm = call.argument("isAlarm");
            setThings(path, downloadedMimeType, isRingt, isNotif, isAlarm);

            result.success(true);
            return;
        } else if (call.method.equals("isWriteGranted")) {
            boolean granted = isSystemWritePermissionGranted();
            result.success(granted);
        }else {
            result.notImplemented();
        }

    }

    @Override
    public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
        channel.setMethodCallHandler(null);
    }
}


lib/ringtone_set_mul.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/services.dart';
import 'package:ringtone_set_mul/src/setter_functions.dart';

/// A class for setting ringtones, notifications, alarms.
class RingtoneSet {
  static const MethodChannel _channel = const MethodChannel('ringtone_set');

  /// Android version.
  static Future<String> get platformVersion async {
    final String version = await _channel.invokeMethod('getPlatformVersion');
    return version;
  }

  /// Android SDK version code.
  static Future<int> get platformSdk async {
    final int sdk = await _channel.invokeMethod('getPlatformSdk');
    return sdk;
  }

  ///多选!!!
  static Future<bool> setMultipleFile(File file,
      {bool isNotification = false,
      bool isRingtone = false,
      bool isAlarm = false}) async {
    return setMultipleFromFile(
        file: file,
        action: 'setMultiple',
        isNotification: isNotification,
        isRingtone: isRingtone,
        isAlarm: isAlarm);
  }

  ///多选
  static Future<bool> setMultipleAsset(String asset,
      {bool isNotification = false,
      bool isRingtone = false,
      bool isAlarm = false}) async {
    return setMultipleFromAsset(
        asset: asset,
        action: 'setMultiple',
        isNotification: isNotification,
        isRingtone: isRingtone,
        isAlarm: isAlarm);
  }

  /// Status of the `WRITE_SETTINGS` permission.
  ///
  /// Returns `true` if the permission has been granted.
  /// On Android 5.1(SDK 22) and older always returns `true`.
  static Future<bool> get isWriteSettingsGranted async {
    final bool granted = await _channel.invokeMethod('isWriteGranted');
    return granted;
  }

  static void get reqSystemPermissions {
    _channel.invokeMethod('reqSystemPermissions');
  }

  /// Sets ringtone from asset.
  static Future<bool> setRingtone(String asset) async {
    return setFromAsset(
      asset: asset,
      action: 'setRingtone',
    );
  }

  /// Sets ringtone from network URL.
  static Future<bool> setRingtoneFromNetwork(String url) async {
    return setFromNetwork(
      url: url,
      action: 'setRingtone',
    );
  }

  /// Sets ringtone from file.
  static Future<bool> setRingtoneFromFile(File file) async {
    return setFromFile(
      file: file,
      action: 'setRingtone',
    );
  }

  /// Sets notification from asset path.
  static Future<bool> setNotification(String asset) async {
    return setFromAsset(
      asset: asset,
      action: 'setNotification',
    );
  }

  /// Sets notification from network URL.
  static Future<bool> setNotificationFromNetwork(String url) async {
    return setFromNetwork(
      url: url,
      action: 'setNotification',
    );
  }

  /// Sets notification from file.
  static Future<bool> setNotificationFromFile(File file) async {
    return setFromFile(
      file: file,
      action: 'setNotification',
    );
  }

  /// Sets alarm from asset path.
  static Future<bool> setAlarm(String asset) async {
    return setFromAsset(
      asset: asset,
      action: 'setAlarm',
    );
  }

  /// Sets alarm from network URL.
  static Future<bool> setAlarmFromNetwork(String url) async {
    return setFromNetwork(
      url: url,
      action: 'setAlarm',
    );
  }

  /// Sets alarm from file.
  static Future<bool> setAlarmFromFile(File file) async {
    return setFromFile(
      file: file,
      action: 'setAlarm',
    );
  }
}
